Containerized the ADAM pipeline from toil-scripts (resolves #139)

fnothaft opened the pull request:
Provides a single node implementation of the ADAM preprocessing pipeline that is
at https://github.com/BD2KGenomics/toil-scripts/tree/master/src/toil_scripts/adam_pipeline.
Depends on https://github.com/BD2KGenomics/toil-scripts/pull/351.

fnothaft committed:
Containerized the ADAM pipeline from toil-scripts (resolves #139)

Provides a single node implementation of the ADAM preprocessing pipeline that is
at https://github.com/BD2KGenomics/toil-scripts/tree/master/src/toil_scripts/adam_pipeline.
Depends on https://github.com/BD2KGenomics/toil-scripts/pull/351.

hannes-ucsc commented on adam-pipeline/Dockerfile:
```diff
@@ -0,0 +1,34 @@
+RUN chmod u+x /usr/local/bin/docker
+
+# Install Toil
+RUN pip install toil
```
pin version?

hannes-ucsc commented on adam-pipeline/test/.gitignore:
```diff
@@ -0,0 +1 @@
+*.processed.bam
```
missing NL at EOL

fnothaft replied:
This is addressed.

hannes-ucsc commented on adam-pipeline/Dockerfile:
```diff
@@ -0,0 +1,34 @@
+
+# Install adam pipeline
+RUN git clone https://github.com/fnothaft/toil-scripts /opt/adam-pipeline
+RUN cd /opt/adam-pipeline && git checkout issues/210-adam-local
```
We mustn't forget that this is temporary and needs to be changed Add GH issue?

fnothaft replied:
Opened #141.

hannes-ucsc commented on adam-pipeline/wrapper.py:
```diff
@@ -0,0 +1,128 @@
+    fp = open(path, 'w')
+
+    # generate config
+    config = '''# ADAM preprocessing pipeline configuration file; autogenerated, do not modify!
```
Consider textwrap.dedent which is part of Python and lets you indent the body of the multi-line string.

hannes-ucsc commented on adam-pipeline/wrapper.py:
```diff
@@ -0,0 +1,128 @@
+def call_pipeline(mount, args):
+    
+    # get uuid and make a work directory
+    uuid = 'Toil-Adam-' + str(uuid4())
```
I'd prefer `toil-adam-`. I hardly ever use upper-case in snake-case identifiers.

hannes-ucsc commented on adam-pipeline/wrapper.py:
```diff
@@ -0,0 +1,128 @@
+    
+    # get uuid and make a work directory
+    uuid = 'Toil-Adam-' + str(uuid4())
+    if not os.path.isdir(mount + uuid):
```
os.path.join

hannes-ucsc replied:
probably a case for extracting a local variable to avoid this kind of bug

hannes-ucsc commented on adam-pipeline/wrapper.py:
```diff
@@ -0,0 +1,128 @@
+
+    # Ensure formatting of command for 2 mount points
+    if len(mounts) == 2:
+        if not all(x['Source'] == x['Destination'] for x in mounts):
```
FYI: I recently added `require()` (proposed by you at some point) to BPL. Not sure if we can assume it being installed as a transitive dependency of toil-scripts. you'd also have to bump the requirement in toil-script's setup.py. Your call.

fnothaft replied:
I've opened #142 for this.

hannes-ucsc commented on adam-pipeline/wrapper.py:
```diff
@@ -0,0 +1,128 @@
+memory: %s
+run-local: true
+local-dir: %s
+''' % (args.known_sites, args.memory, mount)
```
Here I would use .format() with named placeholders. Another trick is 

``` py
>>> x="foo"
>>> "{x}".format(**locals())
'foo'
# dot operator works too
>>> "{x} is {x.__class__}".format(**locals())
"foo is <type 'str'>"
```

fnothaft replied:
Unless you've got strong preferences, I prefer this format. LMK.

hannes-ucsc commented on adam-pipeline/dockstore.cwl:
```diff
@@ -0,0 +1,54 @@
+
+  - id: "#memory"
+	type: string
+	description: "Memory in gigabytes to allocate for this pipeline."
```
GiB or GB?

hannes-ucsc commented on adam-pipeline/README.md:
```diff
@@ -0,0 +1,67 @@
+This pipeline requires a host with Docker 1.9.0 installed. Other versions of
+Docker will soon be supported.
+
+## Genomic tool containers
```
I would remove this section. The commit SHA below is bound to get stale, rendering the WAR ineffective. Can we automate the WAR?

hannes-ucsc commented on adam-pipeline/README.md:
```diff
@@ -0,0 +1,67 @@
+Mirror the absolute path to the parent directory when using Docker's -v mount
+command. Toil's job store and temporary directories will be created inside this
+mount point. "-v /var/run/docker.sock:/var/run/docker.sock" must always be
+supplied. This is necessary for the ADAM pipeline docker container to run
```
Docker

hannes-ucsc commented on adam-pipeline/README.md:
```diff
@@ -0,0 +1,67 @@
+command. Toil's job store and temporary directories will be created inside this
+mount point. "-v /var/run/docker.sock:/var/run/docker.sock" must always be
+supplied. This is necessary for the ADAM pipeline docker container to run
+the docker containers it relies on.
```
Docker

hannes-ucsc commented on adam-pipeline/test.py:
```diff
@@ -0,0 +1,39 @@
+        out = subprocess.check_output(base + tool)
+        self.assertTrue('Please see the complete documentation' in out)
+
+        # run full command on sample inputs and check for existance of output file
```
existence

hannes-ucsc commented:
@fnothaft done with review

fnothaft commented:
Thanks @hannes-ucsc for the quick review! I will address these now.

Some of these issues are duplicated between here and `rnaseq-cgl-pipeline`. I'm going to punt some of those into a follow on ticket, in the interest of time.

fnothaft committed:
Addressing review comments from @hannes-ucsc on PR #140 that touch both adam- and rnaseq-cgl- pipelines.

fnothaft commented:
Addressed review comments. A few of the comments touched code that the `adam-pipeline` copied from the `rnaseq-cgl-pipeline`. I've split those into a separate commit (8e6358f). LMK if you want that squashed down.
